src/aws.ts

```diff
@@ -504,6 +507,237 @@ export namespace $AWS {
+                    });
+                  });
+                },
+              });
+            },
+          }
+        );
+      },
+    }
+  );
+}
+
+type AWSServiceClass = { new (): AWSService };
+
+/**
+ * First we have to extract the names of all Services in the v2 AWS namespace
+ *
+ * @returns "AccessAnalyzer" | "Account" | ... | "XRay"
+ */
+type ServiceKeys = {
+  [K in keyof typeof AWS]: typeof AWS[K] extends AWSServiceClass ? K : never;
+}[keyof typeof AWS];
+
+type StepFunctionSDK = {
+  [serviceName in ServiceKeys]: typeof AWS[serviceName] extends new (
+    ...args: any[]
+  ) => infer Client
+    ? {
+        [methodName in keyof Client]: StepFunctionMethod<Client[methodName]>;
+      }
+    : never;
+};
+
+interface IamOptions {
+  /**
+   * The resources for the IAM statement that will be added to the state machine
+   * role's policy to allow the state machine to make the API call.
+   *
+   * Use `["*"]` to grant access to all resources (discouraged).
+   *
+   * @example ["arn:aws:s3:::my_bucket"]
+   * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_resource.html
+   */
+  iamResources: string[];
+
+  /**
+   * The action for the IAM statement that will be added to the state machine role's
+   * policy to allow the state machine to make the API call.
+   *
+   * By default the action is inferred from the API call (e.g. `$AWS.SDK.CloudWatch.describeAlarms({})` results in `cloudwatch:DescribeAlarms`)
+   *
+   * Use in the case where the IAM action name does not match with the API service/action name
+   * e.g. `s3:ListBuckets` requires `s3:ListAllMyBuckets`.
+   *
+   * @default service:method
+   * @example s3:ListAllMyBuckets
+   * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_action.html
+   */
+  iamActions?: string[];
+
+  /**
+   * The iam conditions to apply to the IAM Statement that will be added to the state machine role's
+   * policy to allow the state machine to make the API call.
+   *
+   * By default no conditions are applied.
+   *
+   * @example
+   * ```
+   * {
+   *   "StringEquals" : { "aws:username" : "johndoe" }
+   * }
+   * ```
+   * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition.html
+   */
+  iamConditions: Record<string, any>;
+}
+
+type StepFunctionMethod<API> = API extends AnyFunction
+  ? Exclude<OverloadUnion<API>, (cb: AnyFunction) => any> extends (
+      input: infer Input extends {}
+    ) => AWS.Request<infer Output, any>
+    ? (input: Input, iamOptions: IamOptions) => Promise<Output>
+    : never
+  : never;
+
+function mapAslSdkServiceName(serviceName: string): string {
+  // source: https://docs.aws.amazon.com/step-functions/latest/dg/supported-services-awssdk.html
+  switch (serviceName) {
+    case "Discovery":
+      return "applicationdiscovery";
+    // TODO: should I simply remove any trailing `Service` or leave the explicit mapping
```

Comment:
+1 to explicit mapping. 

Also would be nice if the user could override this mapping as an escape catch

```ts
$AWS.SDK.SomeService.operation(input, { iam: ..., serviceNameOverride: "the right service name" });
```